tablecodec/tablecodec.go

```diff
@@ -964,15 +1071,95 @@ func GenIndexKey(sc *stmtctx.StatementContext, tblInfo *model.TableInfo, idxInfo
 	return
 }
 
-// GenIndexValue creates encoded index value and returns the result, only support local index
-func GenIndexValue(sc *stmtctx.StatementContext, tblInfo *model.TableInfo, idxInfo *model.IndexInfo, containNonBinaryString bool,
-	distinct bool, untouched bool, indexedValues []types.Datum, h kv.Handle) ([]byte, error) {
-	return GenIndexValueNew(sc, tblInfo, idxInfo, containNonBinaryString, distinct, untouched, indexedValues, h, 0)
+// GenIndexValuePortal is the portal for generating index value.
+func GenIndexValuePortal(sc *stmtctx.StatementContext, tblInfo *model.TableInfo, idxInfo *model.IndexInfo, needRestoredData bool, distinct bool, untouched bool, indexedValues []types.Datum, h kv.Handle, partitionID int64, restoredData []types.Datum) ([]byte, error) {
+	if tblInfo.IsCommonHandle && tblInfo.CommonHandleVersion == 1 {
+		return GenIndexValueNewForClusteredIndexVersion1(sc, tblInfo, idxInfo, needRestoredData, distinct, untouched, indexedValues, h, partitionID, restoredData)
+	}
+	return genIndexValueNew(sc, tblInfo, idxInfo, needRestoredData, distinct, untouched, indexedValues, h, partitionID)
+}
+
+// TryGetCommonPkColumnRestoredIds get the IDs of primary key columns which need restored data if the table has common handle.
+// Caller need to make sure the table has common handle.
+func TryGetCommonPkColumnRestoredIds(tbl *model.TableInfo) []int64 {
+	var pkColIds []int64
+	var pkIdx *model.IndexInfo
+	for _, idx := range tbl.Indices {
+		if idx.Primary {
+			pkIdx = idx
+			break
+		}
+	}
+	if pkIdx == nil {
+		return pkColIds
+	}
+	for _, idxCol := range pkIdx.Columns {
+		if types.NeedRestoredData(&tbl.Columns[idxCol.Offset].FieldType) {
+			pkColIds = append(pkColIds, tbl.Columns[idxCol.Offset].ID)
+		}
+	}
+	return pkColIds
+}
+
+// GenIndexValueNewForClusteredIndexVersion1 generates the index value for the clustered index with version 1(New in v5.0.0).
+func GenIndexValueNewForClusteredIndexVersion1(sc *stmtctx.StatementContext, tblInfo *model.TableInfo, idxInfo *model.IndexInfo, IdxValNeedRestoredData bool, distinct bool, untouched bool, indexedValues []types.Datum, h kv.Handle, partitionID int64, handleRestoredData []types.Datum) ([]byte, error) {
```

Comment:
We can remove the `New` in the name since we have `Version1`.

Reply:
Yeah